File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Query/Filters/PriceRangeFilterDecorator.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.Query.Filters;

/// <summary>
///     A filter that filters inventory items by a price range.
/// </summary>
public sealed class PriceRangeFilterDecorator : InventoryQueryDecoratorBase
{
    private readonly decimal? _max;
    private readonly decimal? _min;

    /// <summary>
    ///     Initializes a new instance of the <see cref="PriceRangeFilterDecorator" /> class.
    /// </summary>
    /// <param name="inner">The inner query to decorate.</param>
    /// <param name="minInclusive">The minimum price (inclusive), or null for no minimum.</param>
    /// <param name="maxInclusive">The maximum price (inclusive), or null for no maximum.</param>
    /// <exception cref="ArgumentOutOfRangeException">Thrown when min or max is negative.</exception>
    /// <exception cref="ArgumentException">Thrown when min is greater than max.</exception>
    public PriceRangeFilterDecorator(IInventoryQuery inner, decimal? minInclusive, decimal? maxInclusive) : base(inner)
    {
        if (minInclusive is not null && minInclusive < 0) throw new ArgumentOutOfRangeException(nameof(minInclusive));
        if (maxInclusive is not null && maxInclusive < 0) throw new ArgumentOutOfRangeException(nameof(maxInclusive));
        if (minInclusive is not null && maxInclusive is not null && minInclusive > maxInclusive)
            throw new ArgumentException("minInclusive cannot be greater than maxInclusive.");

        _min = minInclusive;
        _max = maxInclusive;
    }

    /// <inheritdoc />
    protected override string? AppliedDescription
    {
        get
        {
            if (_min is null && _max is null) return null;
            if (_min is not null && _max is not null)
                return $"Filter: Price between {_min.Value:C} and {_max.Value:C} (inclusive)";
            if (_min is not null)
                return $"Filter: Price >= {_min.Value:C}";
            return $"Filter: Price <= {_max!.Value:C}";
        }
    }

    /// <inheritdoc />
    protected override IReadOnlyList<InventoryItem> Decorate(IReadOnlyList<InventoryItem> items)
    {
        IEnumerable<InventoryItem> filtered = items;

        if (_min is not null) filtered = filtered.Where(i => i.Price >= _min.Value);
        if (_max is not null) filtered = filtered.Where(i => i.Price <= _max.Value);

        return filtered.ToList().AsReadOnly();
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventoryquery/filters/PriceRangeFilterDecorator.java
package assignment2solution.domain.inventoryquery.filters;

import assignment2solution.domain.inventory.InventoryItem;
import assignment2solution.domain.inventoryquery.IInventoryQuery;
import assignment2solution.domain.inventoryquery.InventoryQueryDecoratorBase;

import java.math.BigDecimal;
import java.text.NumberFormat;
import java.util.List;
import java.util.Locale;
import java.util.stream.Collectors;

/**
 * A filter that filters inventory items by a price range.
 */
public final class PriceRangeFilterDecorator extends InventoryQueryDecoratorBase {
    private final BigDecimal min;
    private final BigDecimal max;
    private final NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(Locale.US);

    /**
     * Initializes a new instance of the {@link PriceRangeFilterDecorator} class.
     */
    public PriceRangeFilterDecorator(IInventoryQuery inner, BigDecimal minInclusive, BigDecimal maxInclusive) {
        super(inner);
        if (minInclusive != null && minInclusive.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("minInclusive must be non-negative");
        }
        if (maxInclusive != null && maxInclusive.compareTo(BigDecimal.ZERO) < 0) {
            throw new IllegalArgumentException("maxInclusive must be non-negative");
        }
        if (minInclusive != null && maxInclusive != null && minInclusive.compareTo(maxInclusive) > 0) {
            throw new IllegalArgumentException("minInclusive cannot be greater than maxInclusive.");
        }
        this.min = minInclusive;
        this.max = maxInclusive;
    }

    @Override
    protected String getAppliedDescription() {
        if (min == null && max == null) {
            return null;
        }
        if (min != null && max != null) {
            return "Filter: Price between " + currencyFormatter.format(min) + " and "
                + currencyFormatter.format(max) + " (inclusive)";
        }
        if (min != null) {
            return "Filter: Price >= " + currencyFormatter.format(min);
        }
        return "Filter: Price <= " + currencyFormatter.format(max);
    }

    @Override
    protected List<InventoryItem> decorate(List<InventoryItem> items) {
        var stream = items.stream();

        if (min != null) {
            stream = stream.filter(i -> i.getPrice().compareTo(min) >= 0);
        }
        if (max != null) {
            stream = stream.filter(i -> i.getPrice().compareTo(max) <= 0);
        }

        return stream.collect(Collectors.toUnmodifiableList());
    }
}
